server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocatorTests.java

```diff
@@ -333,6 +336,73 @@ public ShardAllocationDecision decideShardAllocation(ShardRouting shard, Routing
+            l.onResponse(null);
+        });
+        var allocator = new ShardsAllocator() {
+            @Override
+            public void allocate(RoutingAllocation allocation) {// TODO attempt to count up-to-date allocations
+                final var dataNodeId = allocation.nodes().getDataNodes().values().iterator().next().getId();
+                final var unassignedIterator = allocation.routingNodes().unassigned().iterator();
+                while (unassignedIterator.hasNext()) {
+                    String indexName = unassignedIterator.next().getIndexName();
+                    allocatedIndices.add(indexName);
+                    unassignedIterator.initialize(dataNodeId, null, 0L, allocation.changes());
+                }
+            }
+
+            @Override
+            public ShardAllocationDecision decideShardAllocation(ShardRouting shard, RoutingAllocation allocation) {
+                throw new AssertionError("only used for allocation explain");
+            }
+        };
+        var desiredBalanceShardsAllocator = new DesiredBalanceShardsAllocator(allocator, threadPool, () -> rerouteService);
+
+        var discoveryNode = createDiscoveryNode();
+        var clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(DiscoveryNodes.builder().add(discoveryNode).localNodeId(discoveryNode.getId()).masterNodeId(discoveryNode.getId()))
+            .build();
+
+        var indexNameGenerator = new AtomicInteger();
+        var listenersCalled = new AtomicInteger();
+
+        var iterations = 1_000;
+        for (int i = 0; i < iterations; i++) {
+            boolean addNewIndex = i == 0 || randomInt(9) == 0;
+            if (addNewIndex) {
+                var indexName = "index-" + indexNameGenerator.incrementAndGet();
+                var indexMetadata = createIndex(indexName);
+                clusterState = ClusterState.builder(clusterState)
+                    .metadata(Metadata.builder(clusterState.metadata()).put(indexMetadata, true))
+                    .routingTable(RoutingTable.builder(clusterState.routingTable()).addAsNew(indexMetadata))
+                    .build();
+                createdIndices.add(indexName);
+            }
+            var indexName = "index-" + indexNameGenerator;
+            desiredBalanceShardsAllocator.allocate(createAllocationFrom(clusterState), ActionListener.wrap(() -> {
+                assertThat("Listener should be called after index is rerouted", reroutedIndices, hasItem(indexName));
+                listenersCalled.incrementAndGet();
+            }));
+        }
+
+        assertBusy(() -> {
+            assertThat(createdIndices.size(), equalTo(indexNameGenerator.get()));
+            assertThat(allocatedIndices.size(), equalTo(indexNameGenerator.get()));
+            assertThat(reroutedIndices.size(), equalTo(indexNameGenerator.get()));
+            assertThat(listenersCalled.get(), equalTo(iterations));
+        });
```

Comment:
And then this should work right?

```suggestion
        assertTrue(listenersCountdown.await(10, TimeUnit.SECONDS));
        assertThat(createdIndices.size(), equalTo(indexNameGenerator.get()));
        assertThat(allocatedIndices.size(), equalTo(indexNameGenerator.get()));
        assertThat(reroutedIndices.size(), equalTo(indexNameGenerator.get()));
        assertThat(listenersCalled.get(), equalTo(iterations));
```